Fix Eslint Warnings

TigleyM opened the pull request:
fixes #1020 

Fixes the following eslint warnings:
-`react/no-array-index-key`
-`no-return-assign`

howardchung commented on src/components/BurgerMenu/BurgerMenu.jsx:
```diff
@@ -28,8 +28,8 @@ export default class BurgerMenu extends React.Component {
-            {this.props.menuItems.map((item, index) => (
-              <MenuItem className={styles.menuItem} key={index} onTouchTap={item.close && this.handleClose}>
+            {this.props.menuItems.map(item => (
+              <MenuItem className={styles.menuItem} key={item.component.to} onTouchTap={item.close && this.handleClose}>
```
could item.component be null?

howardchung commented on src/components/Match/Overview/Timeline.jsx:
```diff
@@ -205,7 +205,7 @@ const Timeline = ({
                             .filter(player => player.player_slot === aegis[obj.key].player_slot)
                             .map(player => (
-                              <section key={i}>
+                              <section key={`${player.slot}_${player.hero_id}`}>
```
player_slot?

howardchung commented on src/components/Match/TeamfightMap/index.jsx:
```diff
@@ -64,7 +64,7 @@ const getTombStyle = position => position.reduce(
   <div>
     {deathPositions.map((position, index) => (
-      <div key={index}>
+      <div key={`${index + index}`}>
```
this doesn't really solve the issue of not using index for keys?

howardchung commented on src/components/Match/matchColumns.jsx:
```diff
@@ -748,7 +748,7 @@ export const unitKillsColumns = [
     displayFn: (row, col, field) =>
       (<div>
-        {Object.keys(field || {}).map((unit, index) => <div key={index}>{`${field[unit]} ${unit}`}</div>)}
+        {Object.keys(field || {}).map(unit => <div key={`${row}_${col}`}>{`${field[unit]} ${unit}`}</div>)}
```
I think row and col are objects here, this might not make a good key

howardchung commented on src/components/Match/matchColumns.jsx:
```diff
@@ -808,8 +808,8 @@ export const cosmeticsColumns = [
-      field.map((cosmetic, i) =>
-        (<div key={i} className={styles.cosmetics} data-tip data-for={`cosmetic_${cosmetic.item_id}`}>
+      field.map(cosmetic =>
+        (<div key={`${row}_${col}`} className={styles.cosmetics} data-tip data-for={`cosmetic_${cosmetic.item_id}`}>
```
same, maybe use cosmetic.item_id?

howardchung commented on src/components/Player/Pages/Counts/Counts.jsx:
```diff
@@ -12,7 +12,7 @@ import styles from './Counts.css';
   <div className={styles.countsContainer}>
     {Object.keys(counts).map((key, index) => (
-      <div key={index} className={styles.countTable}>
+      <div key={`${index + index}`} className={styles.countTable}>
```
maybe just use key here

howardchung commented on src/components/TabBar/TabBar.jsx:
```diff
@@ -8,7 +8,7 @@ const TabBar = ({ tabs, info, mediaQClass = null }) => (
       {tabs.map((tab, index) => (
         <Link
-          key={index}
+          key={`${index + index}`}
```
use tab.key?

howardchung commented on src/components/Table/Table.jsx:
```diff
@@ -136,7 +136,7 @@ class Table extends React.Component {
               <MaterialTableBody displayRowCheckbox={false} selectable={false}>
                 {data.map((row, index) => (
-                  <MaterialTableRow key={index}>
+                  <MaterialTableRow key={`${currentPage + index}`}>
```
split these with `_`?

howardchung commented on src/components/Header/Header.jsx:
```diff
@@ -40,10 +40,12 @@ const navbarPages = [
-  ...navbarPages.map(item => ({
+  ...navbarPages.map((item, index) => ({
     component: item,
+    key: item.props.to || index,
```
we should avoid using index here?

Also if item.props is undefined item.props.to will crash the application

howardchung replied:
check the entire structure:
item && item.props && items.props.to

Although we'll have to put some unique key value here if any of those are undefined/null

TigleyM replied:
perhaps we can use `item.key` ?

howardchung replied:
oh yeah that's probably better

howardchung commented on src/components/Match/matchColumns.jsx:
```diff
@@ -808,8 +808,8 @@ export const cosmeticsColumns = [
-      field.map((cosmetic, i) =>
-        (<div key={i} className={styles.cosmetics} data-tip data-for={`cosmetic_${cosmetic.item_id}`}>
+      field.map(cosmetic =>
+        (<div key={cosmetic.item_id} className={styles.cosmetics} data-tip data-for={`cosmetic_${cosmetic.item_id}`}>
```
might be possible for the same cosmetic to show up twice (wards?)

howardchung commented on src/components/Table/Table.jsx:
```diff
@@ -204,15 +204,15 @@ class Table extends React.Component {
                       }
                       return (
-                        <MaterialTableRowColumn key={`${index}_${colIndex}`} style={style}>
+                        <MaterialTableRowColumn key={`${index + colIndex}`} style={style}>
```
underscore concatenation probably better

howardchung replied:
comment still applies

howardchung commented on src/components/Table/Table.jsx:
```diff
@@ -204,15 +204,15 @@ class Table extends React.Component {
                 ))}
                 {summable && <MaterialTableRow>
-                  {columns.map((column, colIndex) => (<MaterialTableRowColumn key={`${colIndex}_sum`} style={{ color: column.color }}>
+                  {columns.map((column, colIndex) => (<MaterialTableRowColumn key={`${colIndex + colIndex}_sum`} style={{ color: column.color }}>
```
doubling colindex doesn't solve the problem here of not using index?

howardchung replied:
comment still applies

howardchung commented on src/components/Table/TableHeader.jsx:
```diff
@@ -5,9 +5,9 @@ import TableHeaderColumn from './TableHeaderColumn';
+    {columns.map(column => (
       <TableHeaderColumn
-        key={index}
+        key={column.field}
```
I think column.field is empty string for some columns, there may be duplicate values.

howardchung replied:
maybe column.field || column.name to have a backup

howardchung commented:
Could you rebase this on the latest master to fix the conflicts?

TigleyM commented:
@howardchung Yup, I can get that rebased tonight. 

And thank you for the feedback! I will be addressing your review requests this weekend as I've been pretty busy with school work.

howardchung commented on src/components/BurgerMenu/BurgerMenu.jsx.orig:
```diff
@@ -0,0 +1,59 @@
+          onRequestChange={open => this.setState({ open })}
+        >
+          <Menu>
+<<<<<<< HEAD
```
this should be manually resolved

howardchung commented on src/components/Explorer/ExplorerFormField.jsx:
```diff
@@ -44,7 +44,7 @@ class ExplorerFormField extends React.Component {
       return (<span style={{ width: fieldWidth }}>
         <DatePicker
-          ref={(ref) => { this.datepicker = ref; return null; }}
+          ref={(ref) => { this.datepicker = ref; return this.datepicker; }}
```
I actually prefer returning null here to indicate specifcally that all this function does is an assignment

howardchung commented on src/components/Explorer/ExplorerFormField.jsx.orig:
```diff
@@ -0,0 +1,111 @@
+    if (isDateField) {
+      return (<span style={{ width: fieldWidth }}>
+        <DatePicker
+<<<<<<< HEAD
```
merge conflict

howardchung commented on src/components/Match/Chat/Chat.jsx:
```diff
@@ -457,7 +457,7 @@ class Chat extends React.Component {
               <li
                 id={index}
-                key={index}
+                key={msg.key}
```
key here is the message text and is not unique.

Actually it's kind of difficult to come up with a unique key here that doesn't use index at all, since the same player may say the same thing in the same second.

TigleyM replied:
I wonder if in cases where finding a unique key is difficult we can use the uuid library to generate keys

howardchung replied:
I don't think that will work well since it'll generate a new key every time the component re-renders

howardchung commented:
Could you also edit .eslintrc.json to remove the line making this error a warning instead?

victorlau1 commented:
Hey. I'll take over this one and fix the linting styles.

howardchung commented:
I fixed some of these myself.  There are still 10 lint warnings left, most of these were trickier to find keys for so I've left them for now.  If you feel like tackling this again, please open a new PR! :)
